### DISCUSSION TITLE
cache에 관한 질문 (3차원 그래픽스)

### DISCUSSION THREAD

**Discussion by Lcat12345** [MyeongWoonJang/cpp-study#73] [Q&A] [answered]
## 궁금한 점이 생긴 부분의 원문
게임 프로그래밍을 위한 3차원 그래픽스 中
`인덱스 삼각형 리스트는 저장 공간의 효율성 제고는 몰론 렌더링 성능 향상에도 도움을 준다. 삼각형 메쉬는 GPU에 의해  렌더링되는데, ..., 메쉬의 정점은 GPU의 정점 처리(vertext processing) 단계에 의해 하나하나 변환된 후 캐쉬(cache)에 저장된다. GPU의 그 다음 단계는 삼각형 단위의 연산을 수행하는데, 한 삼각형을 구성하는 세 정점을 먼저 캐쉬에서 찾아보고, 캐쉬에 정점이 없을 경우에만 그 정점을 정점 버퍼에서 꺼내어 정점 처리 단계를 통해 변환한다. 삼각형의 정점을 캐쉬에서 찾을 때 정점의 인덱스가 사용된다.`

## Q1.
정점 버퍼는 배열입니다.  정점을 배열에서 꺼내서 변환한다. 
이말은 배열에서 값을 가지고와서 파이프 라인을 따라 월드 변환, 뷰 변환 등을 하는 것입니다.

`변환된 후 캐쉬(cache)에 저장된다`

이말에서 변환은 월드 변환, 뷰 변환을 말하는 건가요?

## Q2.
cache도 저장소인데 그럼 캐쉬에 변환된 정점이 저장되었으면 따로 GPU(혹은 CPU?)에서 자동으로 저장소를 생성한 건가요.

## Q3.
정점 버퍼가 아닌 캐쉬에서 삼각형 단위로 정점을 찾는 이유는 속도 때문이겟죠?

**Comment by Woon-2** [comment, marked as answer]
1
변환이 아니라 `vertex processing`에 주목하셔야 합니다.   
해당 단계에서 무슨 변환을 하는지는 안 정해져있죠.   
`GPU`가 여러 `vertex`들을 병렬적으로 처리하면서 지역성을 고려해 `cache`를 채울 것입니다.    
아무리 병렬적이어도 10만 개의 `vertex`를 동시에 처리할 순 없잖아요?   
대충 100개씩 동시에 처리가 가능하다 하면, 처리한 `vertex`와 그 `vertex`와 인접한 메모리의 `vertex`들을 합쳐 1000개 정도의 `vertex` 데이터가 `cache`에 저장되겠죠.

2
`GPU` 연산이니 `GPU` `cache`를 이용합니다.   
`CPU` `cache`는 이용하지 않습니다.   

인덱스가 가리키는 정점이 이미 캐시에 있다면 캐시에서 꺼내오고,   
없다면 메모리에서 꺼내옵니다.

한번 정점을 꺼내올 때마다 공간 지역성을 고려해 주변의 정점들을 한꺼번에 꺼내와서,   
시간 지역성을 고려해 기존에 저장된 캐시 메모리를 대체하기 때문에   

지역성이 향상되도록 인덱스의 순서를 배열 내에서 잘 배치하면,   
`cache`의 효율이 올라갑니다.   
모델링 프로그램에서 모델을 저장한 `.obj` 파일을 열어보았을 때,   
인덱스들의 순서가 뒤죽박죽이라고 생각되지 않았나요?   
그것이 아마, `cache` 효율을 최적화하기 위해 배치된 형태일 겁니다.

3
네 그렇습니다.   

<br>
`cache`가 개략적으로 어떻게 동작하는지 알지 못하면 무슨 말인지 이해하기 힘들 겁니다.   

그런 경우 컴퓨터 구조에서 `cache` 부분의 내용을 학습하시는 것을 추천드립니다.

**Comment by Lcat12345** [comment]
도움이 되었습니다. 답변감사합니다.
